File: scall/src/main/resources/prolog/parsing_table.pl
% this file provides the theory used to compute the parsing table starting from
% a knowledge base consisting of declared productions, declared terminals, a start symbol.
% the following functors are to be used in the knowledge base:
% start_symbol(Nonterminal).
% production(Head, Body).
% terminal(Terminal).

% first_str(String, Entry)
% relates a string of symbols with an entry in its FIRST set
% the ε entry (empty string) is represented with the number 0
first_str([], 0).
first_str([A | _], A) :- terminal(A).
first_str([X | _], A) :- first(X, A), A \== 0.
first_str([X | T], A) :- first(X, 0), first_str(T, A).

% first(Nonterminal, Entry)
% relates a nonterminal with an entry in its FIRST set
% the ε entry is represented with the number 0
first(X, A) :- production(X, B), first_str(B, A).

% following(Nonterminal, String, Head)
% relates a nonterminal with a string of symbols that follows it within a production body
% along with the head of that same production
following(X, S, H) :- production(H, B), append(_, [X | S], B), \+ terminal(X).

% follow(Nonterminal, Entry)
% relates a nonterminal with an entry in its FOLLOW set
% the $ entry (end of input) is represented with the number 1
follow(X, 1) :- start_symbol(X).
follow(X, A) :- following(X, S, _), first_str(S, A), A \== 0.
follow(X, A) :- following(X, S, H), H \== X, first_str(S, 0), follow(H, A).

% parsing_cell(Nonterminal, Terminal, Body)
% relates a position within the parsing table, given by a nonterminal along with a terminal
% (possibly end of input), with the value in that cell, being a production body
% assuming the starting grammar is LL(1), each cell will have only one possible value
parsing_cell(X, A, B) :- production(X, B), first_str(B, A), A \== 0.
parsing_cell(X, A, B) :- production(X, B), first_str(B, 0), follow(X, A).

File: scall/src/main/scala/parser/ParsingTable.scala
package it.unibo.scall
package parser

import grammar.Element.{TerminalOrEoi, Eoi}
import grammar.Element.Terminal
import grammar.ProcessedGrammar
import grammar.ProcessedGrammar.{AnyNonterminal, AnySymbol, SymbolSeq}
import util.CollectionUtils.mapEntries
import util.Scala2P
import util.Scala2P.{*, given}

/** Representation of an LL(1) [[ParsingTable parsing table]].
 *  Cells are identified by a pair consisting of a nonterminal, which is the current parsing state,
 *  and a terminal, which is read from input, or possibly an [[Eoi end of input]].
 *  Values within the table are production bodies.
 *  A parsing table can be computed from a [[ProcessedGrammar]] by means of the [[compute]] method,
 *  which makes use of a [[Scala2P tuProlog engine]] and the theory from `/prolog/parsing_table.pl`.
 */
object ParsingTable:

  type ParsingTable = Map[(AnyNonterminal, TerminalOrEoi), SymbolSeq]

  private val theoryFile = "/prolog/parsing_table.pl"

  private lazy val engine = engineWithTheoryFile(
    getClass.getResourceAsStream(theoryFile)
  )
  given Scala2P = engine

  /** Computes a parsing table from the given grammar.
   *  The algorithm is implemented in prolog (see `/prolog/parsing_table.pl`),
   *  by computing the FIRST and FOLLOW sets for each nonterminal and production body.
   *  If the supplied grammar is not LL(1), the result is undefined.
   */
  def compute(grammar: ProcessedGrammar): ParsingTable =
    registerScope:
      withKnowledge(grammarKnowledge(grammar)): () =>
        val X = variable("X"); val A = variable("A"); val B = variable("B")
        val parsingTableGoal = compoundTerm("parsing_table_cell", X, A, B)
        parsingTableGoal.solveAll.collectSuccess: s =>
          val row = s.getRegistered[AnyNonterminal](X)
          val col = s.get(A):
            case Registered[Terminal](t) => t
            case Int(1) => Eoi
          val value = s.getRegisteredList[AnySymbol](B)
          (row, col) -> value
        .toMap

  private def grammarKnowledge(g: ProcessedGrammar)(using scope: RegisterScope) =
    import scala.language.implicitConversions
    given TermConversion[AnySymbol] = register
    val t = g.terminals.filter(!_.isSkipped).map: t =>
      compoundTerm("terminal", t)
    val p = g.productions.mapEntries: (head, body) =>
      compoundTerm("production", head, body)
    val s = compoundTerm("start_symbol", g.startSymbol)
    t ++ p :+ s
